Add new topic on deploying OTeL for IMM

tmorrisonsfx opened the pull request:
Jira issue: https://signalfuse.atlassian.net/browse/OTL-335

What changed?

- Created a new topic that explains how to deploy OTeL Collector for IMM 
- Updated existing Smart Agent topics to reference the new OTeL topic (see https://github.com/signalfx/product-docs/pull/777)
- Updated the integrations index to reference the new OTeL topic (see https://github.com/signalfx/product-docs/pull/777)

tmorrisonsfx commented:
@brianashby-sfx @flands PTAL.

brianashby-sfx commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,110 @@
+- Use the [quick installation](https://docs.signalfx.com/en/latest/integrations/agent/quick-install.html "Quick Install") option for simplified SignalFx Smart Agent command-line installation on a single host.
+- Use the [advanced installation](https://docs.signalfx.com/en/latest/integrations/agent/advanced-install-options.html "Advanced Installation Options") option for bulk deployments and for configuring various monitors for your environment.
+
+**Step 2: Configure the SignalFx Smart Agent**
```
we need to address what it looks like to send data to an otel collector, e.g., including the hostname:port of the otel collector to export data to the collector. 

even though we link to an example config on line 34, we could link to this for additional example configs: https://github.com/signalfx/signalfx-agent/tree/master/deployments.

brianashby-sfx commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,110 @@
+  
+See [Configuration](https://opentelemetry.io/docs/collector/configuration/ "OpenTelemetry Collector Configuration") for more information about the OpenTelemetry Collector components.
+
+### Deploy an OpenTelemetry Collector
```
this seems like a redundant heading because of the H2 on line 36

brianashby-sfx commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,110 @@
+To deploy an OpenTelemetry Collector:
+
+1. Download the [latest release](https://github.com/open-telemetry/opentelemetry-collector-contrib/releases "OpenTelemetry Collector contributions releases") of the OpenTelemetry Collector contributions version from GitHub.
+2. Create a configuration file that defines components for the OpenTelemetry Collector. A sample configuration file is provided below.
```
we may be able to use this directory to reference additional examples: https://github.com/open-telemetry/opentelemetry-collector/tree/master/examples.

also, consider removing detailed config info about things like `memory_limiter` that may be documented in another location and is subject to change without our knowledge or oversight. we could instead link to existing otel resources.

tmorrisonsfx replied:
Can you provide more guidance on what to remove from the config? For example, should the `memory_limiter` be mentioned in the config without the specific options (`ballast_size_mib`, `check_interval`, etc.)?

brianashby-sfx replied:
i'm thinking we could remove lines 85-92. information about `memory_limiter` is available at https://github.com/open-telemetry/opentelemetry-collector/tree/master/processor/memorylimiter, so we may not want to document it in comments here.

tmorrisonsfx replied:
Per our Slack convo, I removed lines 72-84.

brianashby-sfx commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,110 @@
+
+## Deploy the SignalFx Smart Agent
+
+The SignalFx platform is built to ingest, store, analyze, visualize, and alert on metrics at scale. A metric is anything that is measurable (you can assign a numerical value to it) and variable (changes over time). The following are examples of metrics:
+
+* CPU utilization % of a server
+* Response time in milliseconds of an API call
+* The number of unique users who logged in over the previous 24-hour period
```
this may be unnecessary for the purpose of the topic.

instead perhaps we can talk about how the smart agent handles receiving metrics from integrations before exporting to the otel collector for central management and processing.

brianashby-sfx commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,110 @@
+
+The SignalFx Smart Agent is configured primarily by a YAML document located at `/etc/signalfx/agent.yaml`. The location of the configuration file can be specified by the `-config` flag to the agent binary (`signalfx-agent`).
+
+See [Agent Configuration](https://docs.signalfx.com/en/latest/integrations/agent/config-schema.html "Agent Configuration") for complete details, including configuration options and an example configuration file.
```
```suggestion
See [Agent Configuration](https://docs.signalfx.com/en/latest/integrations/agent/config-schema.html) for complete details, including configuration options and an example configuration file.
```

brianashby-sfx commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,110 @@
+| **Component** | **Name**        | **Description**                                                                                       |  
+|---------------| :---------------:  |-------------------------------------------------------------------------------------------------------|
+| Receiver      | `signal-fx`     | Component that sets the endpoint for receiving metrics data with the SignalFx metric data format.     |      
+| Processor     | Various         | Component that pre-processes data before it is exported.                                              |      
```
can we be more specific about any processor required for a minimal config?

brianashby-sfx commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,110 @@
+
+After installing the SignalFx Smart Agent on each host, you can deploy an OpenTelemetry Collector in each datacenter/region/cluster where traced applications run. In general, an OpenTelemetry Collector should receive data from the SignalFx Smart Agent.
+
+The OpenTelemetry Collector uses pipelines to receive, process, and export trace data with components conveniently known as receivers, processors, and exporters. Set up pipelines with services. You can also add extensions that provide an OpenTelemetry Collector with additional functionality, such as diagnostics and health checks. The OpenTelemetry Collector has two versions: a [core version](http ...
```
it may be helpful to provide example use cases that explain why someone would deploy an otel collector between a smart agent and sfx ingest endpoint.

tmorrisonsfx replied:
I pulled some information from this specs page on use cases: https://confluence.splunk.com/display/PROD/SFx+Deliverables

brianashby-sfx commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,110 @@
+# Deploy the SignalFx Smart Agent and OpenTelemetry Collector for Infrastructure Monitoring
```
```suggestion
# Deploy the SignalFx Smart Agent and OpenTelemetry Collector for SignalFx Infrastructure Monitoring
```

brianashby-sfx commented:
@flands could you pay special attention to the example config?

tmorrisonsfx commented:
@brianashby-sfx Thanks for the review! I incorporated most of your feedback. I have one follow-up question on the config example.

tmorrisonsfx commented:
@flands @brianashby-sfx PTAL.

flands commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,101 @@
+
+SignalFx Infrastructure Monitoring helps you gather metrics related to your system's performance. Metrics provide information about processes running inside the system, including counters, cumulative counters, and gauges.
+
+OpenTelemetry provides the libraries, agents, and other components that you need to capture telemetry from your services so that you can better observe, manage, and debug them. The following are just a few scenarios where you can use OpenTelemetry to monitor your solutions:
+
+* Creating custom async libraries 
+* Instrumenting metrics
+* Using multiple context propagation formats (B3, W3C) in parallel
+
+The OpenTelemetry Collector is a stand-alone service that can ingest metrics from various sources. The OpenTelemetry collector provides additional flexibility in configuration options. You can now configure integrations to send data to an OpenTelemetry Collector to centrally manage data sent to SignalFx.
```
Not sure I would cover this here -- you are only talking about the collector below, so instrumentation is separate. I would likely skip all this but anything specific to Otel should be covered in Otel section below

flands commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,101 @@
+
+The OpenTelemetry Collector is a stand-alone service that can ingest metrics from various sources. The OpenTelemetry collector provides additional flexibility in configuration options. You can now configure integrations to send data to an OpenTelemetry Collector to centrally manage data sent to SignalFx.
+
+Deploying the SignalFx Smart Agent and OpenTelemetry Collector is a two-step process:
+
+1. [Deploy the SignalFx Smart Agent](#deploy-the-signalfx-smart-agent)
+2. [Deploy an OpenTelemetry Collector for SignalFx](#deploy-an-opentelemetry-collector-for-signalfx)
```
The happy path is app > smart agent.> ingest (no otel) so need to cover when you deploy otel and why -- a picture would help here

flands commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,101 @@
+
+## Deploy an OpenTelemetry Collector for SignalFx
+
+After installing the SignalFx Smart Agent on each host, you can deploy an OpenTelemetry Collector in each datacenter/region/cluster where traced applications run. In general, an OpenTelemetry Collector should receive data from the SignalFx Smart Agent.
```
Need to add the why -- reasons are the same as the gateway today

flands replied:
You would likely want a cluster of collectors behind a load balancer to have HA

flands commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,101 @@
+
+After installing the SignalFx Smart Agent on each host, you can deploy an OpenTelemetry Collector in each datacenter/region/cluster where traced applications run. In general, an OpenTelemetry Collector should receive data from the SignalFx Smart Agent.
+
+The OpenTelemetry Collector uses pipelines to receive, process, and export trace data with components conveniently known as receivers, processors, and exporters. Set up pipelines with services. You can also add extensions that provide an OpenTelemetry Collector with additional functionality, such as diagnostics and health checks. The OpenTelemetry Collector has two versions: a [core version](http ...
```
I am not sure if the customer cares -- if they do, this info is already covered in the Otel repo so perhaps just link so you do not need to keep updating this

flands commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,101 @@
+
+The OpenTelemetry Collector uses pipelines to receive, process, and export trace data with components conveniently known as receivers, processors, and exporters. Set up pipelines with services. You can also add extensions that provide an OpenTelemetry Collector with additional functionality, such as diagnostics and health checks. The OpenTelemetry Collector has two versions: a [core version](http ...
+
+SignalFx uses the components described in the following table to send data to an OpenTelemetry Collector and to receive data from an OpenTelemetry Collector:
+
+| **Component** | **Name**        | **Description**                                                                                       |  
+|---------------| :---------------:  |-------------------------------------------------------------------------------------------------------|
+| Receiver      | `signal-fx`     | Component that sets the endpoint for receiving metrics data with the SignalFx metric data format.     |      
+| Processor     | `memory_limiter`, `batch`, and `queued_retry` are the recommended processors.         | Component that pre-processes data before it is exported.                                              |      
+| Exporter      | `signal-fx`     | Component that forwards data to SignalFx with the metric data format.                                 |
```
Again, not sure the customer cares -- they need configuration, but likely do not need the specifics of things. Also missing processors here

flands commented on docs/imm-otel.md:
````diff
@@ -0,0 +1,101 @@
+```
+# Save this file as collector.yaml.
+# Extensions are optional and are provided to monitor
+# the health of the OpenTelemetry Collector.  
+    extensions:
+      health_check: {}
+      pprof: {}
+      zpages: {}
+# The following section is used to collect the
+# OpenTelemetry Collector metrics.
+    receivers:
+      signalfx:
+# The SignalFx receiver accepts metrics in the SignalFx
+# proto format. This allows the collector to receive
+# metrics from the OpenTelemetry Collector.
+    processors:
+# Processors are not enabled by default; however, the following
+# processors are recommended.
+    batch:
+# The batch processor must be defined in the pipeline after the
+# memory_limiter as well as any sampling processors.
+    memory_limiter:
+      ballast_size_mib: 683
+      check_interval: 2s
+      limit_mib: 1800
+      spike_limit_mib: 500
+ # Enabling the memory_limiter is strongly recommended for every pipeline. 
+    exporters:
+# One or more exporters must be configured.
+# Metrics
+     signalfx:
+        access_token: "YOUR_ACCESS_TOKEN"
+        realm: "YOUR_SIGNALFX_REALM"
+     service:
+       pipelines:
+         metrics:
+           receivers: [signalfx]
+           processors: [memory_limiter, batch]
+           exporters: [signalfx]
+       extensions: [health_check, zpages, pprof]
````
For APM we put the config on github: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/sapmexporter/examples/signalfx-collector.yaml should do the same for IMM so not in the docs and do not get out of sync. I am happy to do this and provide a link

tmorrisonsfx replied:
Thanks. Can you provide the link to the YAML file?

flands replied:
same link above works for IMM

flands commented on docs/imm-otel.md:
````diff
@@ -0,0 +1,101 @@
+```
+3. Deploy the Open Telemetry Collector.
+```
+otelcontribcol --config collector.yaml --mem-ballast-size-mib=683
````
The ballast depends on the amount of memory allocated to the collector -- need some note on this

tmorrisonsfx replied:
Where can I find this information?

flands replied:
https://github.com/open-telemetry/opentelemetry-collector/tree/master/processor/memorylimiter

flands commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,101 @@
+# Deploy the SignalFx Smart Agent and OpenTelemetry Collector for SignalFx Infrastructure Monitoring
```
```suggestion
# Deploy Data Collection for SignalFx Infrastructure Monitoring
```

flands commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,101 @@
+
+SignalFx Infrastructure Monitoring helps you gather metrics related to your system's performance. Metrics provide information about processes running inside the system, including counters, cumulative counters, and gauges.
+
+OpenTelemetry provides the libraries, agents, and other components that you need to capture telemetry from your services so that you can better observe, manage, and debug them. The following are just a few scenarios where you can use OpenTelemetry to monitor your solutions:
+
+* Creating custom async libraries 
+* Instrumenting metrics
+* Using multiple context propagation formats (B3, W3C) in parallel
+
```
```suggestion
```

flands commented on docs/imm-otel.md:
```diff
@@ -0,0 +1,101 @@
+
+See [Agent Configuration](https://docs.signalfx.com/en/latest/integrations/agent/config-schema.html "Agent Configuration") for configuration options and an example configuration file. See [deployments](https://github.com/signalfx/signalfx-agent/tree/master/deployments "Deployments") for more information on the various ways to deploy the SignalFx Smart Agent.
+
+## Deploy an OpenTelemetry Collector for SignalFx
```
```suggestion
## Deploy the OpenTelemetry Collector
```

flands commented on docs/imm-otel.md:
````diff
@@ -0,0 +1,101 @@
+
+1. Download the [latest release](https://github.com/open-telemetry/opentelemetry-collector-contrib/releases "OpenTelemetry Collector contributions releases") of the OpenTelemetry Collector contributions version from GitHub.
+2. Create a configuration file that defines components for the OpenTelemetry Collector. A sample configuration file is provided below. See [examples](https://github.com/open-telemetry/opentelemetry-collector/tree/master/examples “OTeL examples”) for an OpenTelemetry Collector demo.
+```
````
I would strongly advise to link to the github config as it will be updated while the docs likely will not

tmorrisonsfx commented:
Closing this PR in favor of https://github.com/signalfx/product-docs/pull/802. The new PR is in the product-docs repo, which will provide more control over when we can release it. I have addressed all comments in this PR in the new PR.
